kong/dns/client.lua

```diff
@@ -0,0 +1,691 @@
+
+local TYPE_TO_NAME = {
+  [TYPE_SRV] = "SRV",
+  [TYPE_A] = "A",
+  [TYPE_AAAA] = "AAAA",
+  [TYPE_A_AAAA] = "A/AAAA",
+}
+
+local HIT_L3 = 3 -- L1 lru, L2 shm, L3 callback, L4 stale
+
+local HIT_LEVEL_TO_NAME = {
+  [1] = "hit_lru",
+  [2] = "hit_shm",
+  [3] = "miss",
+  [4] = "hit_stale",
+}
+
+-- client specific error
+local CACHE_ONLY_ERROR_CODE = 100
+local CACHE_ONLY_ERROR_MESSAGE = "cache only lookup failed"
+local CACHE_ONLY_ANSWERS = {
+  errcode = CACHE_ONLY_ERROR_CODE,
+  errstr = CACHE_ONLY_ERROR_MESSAGE,
+}
+
+local EMPTY_RECORD_ERROR_CODE = 101
+local EMPTY_RECORD_ERROR_MESSAGE = "empty record received"
+
+
+-- APIs
+
+local _M = {
+  TYPE_SRV = TYPE_SRV,
+  TYPE_A = TYPE_A,
+  TYPE_AAAA = TYPE_AAAA,
+}
+local MT = { __index = _M }
+
+
+local TRIES_MT = { __tostring = cjson.encode, }
+
+
+local function stats_init_name(stats, name)
+  if not stats[name] then
+    stats[name] = {}
+  end
+end
+
+
+local function stats_increment(stats, name, key)
+  stats[name][key] = (stats[name][key] or 0) + 1
+end
+
+
+local function stats_set_count(stats, name, key, value)
+  stats[name][key] = value
+end
+
+
+local init_hosts do
+  local function insert_answer_into_cache(cache, hosts_cache, address, name, qtype)
+    local answers = {
+      ttl = LONG_LASTING_TTL,
+      expire = now() + LONG_LASTING_TTL,
+      {
+        name = name,
+        type = qtype,
+        address = address,
+        class = 1,
+        ttl = LONG_LASTING_TTL,
+      },
+    }
+
+    hosts_cache[name .. ":" .. qtype] = answers
+    hosts_cache[name .. ":" .. TYPE_A_AAAA] = answers
+  end
+
+  -- insert hosts into cache
+  function init_hosts(cache, path)
+    local hosts = parse_hosts(path)
+    local hosts_cache = {}
+
+    for name, address in pairs(hosts) do
+      name = string_lower(name)
+
+      if address.ipv6 then
+        insert_answer_into_cache(cache, hosts_cache, address.ipv6, name, TYPE_AAAA)
+      end
+
+      if address.ipv4 then
+        insert_answer_into_cache(cache, hosts_cache, address.ipv4, name, TYPE_A)
+      end
+    end
+
+    return hosts, hosts_cache
+  end
+end
+
+
+-- distinguish the worker_events sources registered by different new() instances
+local ipc_counter = 0
+
+function _M.new(opts)
+  opts = opts or {}
+
+  local enable_ipv4, enable_ipv6, enable_srv
+
+  for i, typstr in ipairs(opts.family or DEFAULT_FAMILY) do
+    typstr = typstr:upper()
+
+    if typstr == "A" then
+      enable_ipv4 = true
+
+    elseif typstr == "AAAA" then
+      enable_ipv6 = true
+
+    elseif typstr == "SRV" then
+      enable_srv = true
+
+    else
+      return nil, "Invalid dns type in dns_family array: " .. typstr
+    end
+  end
+
+  log(NOTICE, PREFIX, "supported types: ", enable_srv and "srv " or "",
+              enable_ipv4 and "ipv4 " or "", enable_ipv6 and "ipv6 " or "")
+
+  -- parse resolv.conf
+  local resolv, err = utils.parse_resolv_conf(opts.resolv_conf, opts.enable_ipv6)
```

Comment:
Why did we not localize `utils.parse_resolv_conf` at the top of module?

Reply:
yeah good catch. will fix it